python/graphstorm/dataloading/sampler.py

```diff
@@ -243,3 +246,144 @@ def _generate(self, g, eids, canonical_etype):
+        A dict of reverse edge type info.
+    """
+    def __init__(
+        self,
+        fanouts,
+        edge_dir="in",
+        prob=None,
+        mask=None,
+        replace=False,
+        prefetch_node_feats=None,
+        prefetch_labels=None,
+        prefetch_edge_feats=None,
+        output_device=None,
+        reverse_edge_types_map=None,
+    ):
+        self.mask = mask
+        self.reverse_edge_types_map = reverse_edge_types_map
+        self.original_edge_types_map = {
+            val: key for key, val in reverse_edge_types_map.items()
+        } if reverse_edge_types_map is not None else None
+
+        super().__init__(
+            fanouts=fanouts,
+            edge_dir=edge_dir,
+            prob=prob,
+            mask=None, # Do neighbor sampling with out edge mask
+            replace=replace,
+            prefetch_node_feats=prefetch_node_feats,
+            prefetch_labels=prefetch_labels,
+            prefetch_edge_feats=prefetch_edge_feats,
+            output_device=output_device
+        )
+
+    def sample_blocks(self, g, seed_nodes, exclude_eids=None):
+        """Generates a list of blocks from the given seed nodes.
+
+        This function must return a triplet where the first element is the input node IDs
+        for the first GNN layer (a tensor or a dict of tensors for heterogeneous graphs),
+        the second element is the output node IDs for the last GNN layer, and the third
+        element is the said list of blocks.
+
+        Parameters
+        ----------
+        g: DGLGraph
+            Graph to sample blocks.
+        seed_nodes: dict of tensors
+            Seed nodes.
+        exclude_eids: func
+            Operations to exlude eids.
+        """
+        output_nodes = seed_nodes
+        blocks = []
+        for fanout in reversed(self.fanouts):
+            frontier = g.sample_neighbors(
+                seed_nodes,
+                fanout,
+                edge_dir=self.edge_dir,
+                prob=self.prob,
+                replace=self.replace,
+                output_device=self.output_device,
+                exclude_edges=exclude_eids,
+            )
+            eid = frontier.edata[EID]
+            new_eid = dict(eid)
+            if self.mask is not None:
+                new_edges = {}
+                for etype in frontier.canonical_etypes:
+                    if self.mask in g.edges[etype].data:
+                        # train mask in data
+                        if etype in eid:
+                            mask = g.edges[etype].data[self.mask][eid[etype]].bool()
+                            new_edges[etype] = mask
+                            new_eid[etype] = eid[etype][mask]
+
+                    elif self.original_edge_types_map is not None and \
+                        self.mask in g.edges[self.original_edge_types_map[etype]].data:
+                        # handle reverse edges
+                        original_etype = self.original_edge_types_map[etype]
```

Comment:
is this naming correctly? is this original etype or reversed etype?

Reply:
It is the reverse map of reverse_edge_types_map.